ra/ra.go

```diff
@@ -1795,8 +1820,293 @@ func (ra *RegistrationAuthorityImpl) revokeCertificate(ctx context.Context, cert
+				return nil, fmt.Errorf("requester does not control all names in cert with serial %q", serialString)
+			}
+		}
+
+		// Override the requested revocation reason to match the request method.
+		req.Code = ocsp.CessationOfOperation
+		logEvent.Reason = req.Code
+	}
+
+	issuerID := issuance.GetIssuerNameID(cert)
+	err = ra.revokeCertificate(
+		ctx,
+		cert.SerialNumber,
+		int64(issuerID),
+		revocation.Reason(req.Code),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	err = ra.purgeOCSPCache(ctx, cert, int64(issuerID))
+	if err != nil {
+		return nil, err
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+// RevokeCertByKey revokes the certificate in question. It always uses
+// reason code 1 (keyCompromise). It ensures that they public key is added to
+// the blocked keys list, even if revocation otherwise fails. It attempts to
+// purge the certificate from the Akamai cache, but it does not hard-fail if
+// doing so is not successful, because the cache will drop the old OCSP response
+// in less than 24 hours anyway.
+func (ra *RegistrationAuthorityImpl) RevokeCertByKey(ctx context.Context, req *rapb.RevokeCertByKeyRequest) (*emptypb.Empty, error) {
+	if req == nil || req.Cert == nil {
+		return nil, errIncompleteGRPCRequest
+	}
+
+	cert, err := x509.ParseCertificate(req.Cert)
+	if err != nil {
+		return nil, err
+	}
+
+	issuerID := issuance.GetIssuerNameID(cert)
+
+	logEvent := certificateRevocationEvent{
+		ID:           core.NewToken(),
+		SerialNumber: core.SerialToString(cert.SerialNumber),
+		Reason:       ocsp.KeyCompromise,
+		Method:       "key",
+		RequesterID:  0,
+	}
+	defer func() {
+		if err != nil {
+			logEvent.Error = err.Error()
+		}
+		ra.log.AuditObject("Revocation request:", logEvent)
+	}()
+
+	// We revoke the cert before adding it to the blocked keys list, to avoid a
+	// race between this and the bad-key-revoker. But we don't check the error on
+	// from this operation until after we add to the blocked keys list, since that
+	// add needs to happen no matter what.
+	revokeErr := ra.revokeCertificate(
+		ctx,
+		cert.SerialNumber,
+		int64(issuerID),
+		revocation.Reason(ocsp.KeyCompromise),
+	)
+
+	// Now add the public key to the blocked keys list, and report the error if
+	// there is one. It's okay to error out here because failing to add the key
+	// to the blocked keys list is a worse failure than failing to revoke in the
+	// first place, because it means that bad-key-revoker won't revoke the cert
+	// anyway.
+	digest, err := core.KeyDigest(cert.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	_, err = ra.SA.AddBlockedKey(ctx, &sapb.AddBlockedKeyRequest{
+		KeyHash: digest[:],
+		Added:   ra.clk.Now().UnixNano(),
+		Source:  "API",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Finally check the error from revocation itself. If there was an error,
+	// try to re-revoke the cert, in case the error was due to it being already
+	// revoked for some reason other than keyCompromise.
+	if revokeErr != nil {
```

Comment:
We should probably be checking for a sentinel error here instead of immediately attempting another revocation if the prior one failed for any reason. This feels especially true when the way that reRevokeCertificate guards against just updating the status is by calling `sa.GetCertificateStatus()`.